src/main/java/com/amazon/opendistroforelasticsearch/knn/index/MethodComponent.java

```diff
@@ -0,0 +1,136 @@
+ */
+public class MethodComponent {
+
+    private String name;
+    private Map<String, Parameter<?>> parameters;
+
+    /**
+     * Constructor
+     *
+     * @param builder to build method component
+     */
+    private MethodComponent(Builder builder) {
+        this.name = builder.name;
+        this.parameters = builder.parameters;
+    }
+
+    /**
+     * Get the name of the component
+     *
+     * @return name
+     */
+    public String getName() {
+        return name;
+    }
+
+    /**
+     * Get the parameters for the component
+     *
+     * @return parameters
+     */
+    public Map<String, Parameter<?>> getParameters() {
+        return parameters;
+    }
+
+    /**
+     * Generate extra parameters for this component that are not contained in method string
+     *
+     * @return map of extra parameters
+     */
+    public Map<String, Object> generateExtraParameterMap(Map<String, Object> inParameters) {
+        if (inParameters == null) {
+            return Collections.emptyMap();
+        }
+        return inParameters.entrySet().stream().filter(v -> !parameters.get(v.getKey()).isInMethodString()).collect(
+                Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+    }
+
+    /**
+     * Validate that the methodComponentContext is a valid configuration for this methodComponent
+     *
+     * @param methodComponentContext to be validated
+     */
+    public void validate(KNNMethodContext.MethodComponentContext methodComponentContext) {
+        Map<String, Object> providedParameters = methodComponentContext.getParameters();
+        if (providedParameters != null) {
+            for (Map.Entry<String, Object> parameter : providedParameters.entrySet()) {
+                if (!parameters.containsKey(parameter.getKey())) {
+                    throw new ValidationException();
+                }
+
+                parameters.get(parameter.getKey()).validate(parameter.getValue());
+            }
+        }
+    }
+
+    /**
+     * Builder class for MethodComponent
+     */
+    public static class Builder {
+
+        private String name;
+        private Map<String, Parameter<?>> parameters;
+
+        /**
+         * Method to get a Builder instance
+         *
+         * @param name of method component
+         * @return Builder instance
+         */
+        public static MethodComponent.Builder builder(String name) {
+            return new MethodComponent.Builder(name);
+        }
+
+        private Builder(String name) {
+            this.name = name;
+            this.parameters = new HashMap<>();
+        }
+
+        /**
+         * Add parameter entry to parameters map
+         *
+         * @param parameterName name of the parameter
+         * @param parameter parameter to be added
+         * @return this builder
+         */
+        public Builder putParameter(String parameterName, Parameter<?> parameter) {
```

Comment:
may be addParameter?

Reply:
My thinking was that since it is adding a parameter into a map, it should be similar to Map.put. What do you think?

Reply:
but parameter map implementation is abstracted by Builder. So from Builder point of you we are adding parameter, but internally we are using map to insert.

Reply:
You are right. I will fix